### PR TITLE
Async/Lazy load photo

### LINKED ISSUE #540: Load photos asynchronous from Azure Blob Storage
Fetch photos regardless of events

### PR DESCRIPTION
**Pull request by dyats**
dev
## GitHub Project

* [Main GitHub Project ticket](https://github.com/EventsExpress/EventsExpress/projects/2)

## Code reviewers

- [x] @iuriikhrystiuk 

### Second Level Review

- [ ] @sand0 

## Summary of issue

Photos was loading with all users data and that was consuming a lot of time

## Summary of change

Added posibility of photo lazy loading
If photo haven't been found than it will be replaced by default photo

## Testing approach

Tested manually

## CHECK LIST
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: EventsExpress/ClientApp/src/components/chat/chat.js [hunk 1]
```diff
@@ -85,9 +87,9 @@ class Chat extends Component {
                             <div className="d-flex bd-highlight">
                                 {sender != null &&
                                     <ButtonBase>
-                                        {sender.photoUrl
-                                            ? <Avatar className='SmallAvatar' src={sender.photoUrl} />
-                                            : <Avatar className='SmallAvatar' >{sender.username.charAt(0).toUpperCase()}</Avatar>}
+                                        <Avatar className='SmallAvatar'
+                                                src={`api/photo/GetUserPhoto?id=${sender.id}`}
+                                                imgProps={{ onError: (e) => { e.target.onerror = null; e.target.src = `${userImage}` } }}/>
```
Comment (sand0) [inline EventsExpress/ClientApp/src/components/chat/chat.js]: I guess we should use our CustomAvatar component here, due to DRY

File: EventsExpress/ClientApp/src/constants/userImage.js [hunk 2]
```diff
@@ -0,0 +1 @@
...nkB2GPOP/2iTCE64rIOOK2OoiB8FPzS0NNyMlERM2Vu4f4AKJeqrY03EA2xVOX0zK84PBzvhX/qeQ75xH1CoIXOQDwX/slsM91+pFTjvPo1BES8qF+dPhfPR7vKgrfVbCEs4tLua7oRj0d7SCZnF1dzh0DEWf5B9icyCW9HI6r9U+kXOTgU2Xr/jvRzmX3xcjR2e3B24lV6A9Gr92d+13oQ3o0KUstRqXKSQGQYSr7eeThejSaktqNJAlH48yf/mCb4UXFqOjL5OkkgtMG/M/1q72F4MSqtQXho6jfcJBC6gt/2zK+6Z+O9yBXVz8gsqX6PsArLdH92MW/3Ipcl+zTdU7ee8LKbmnXZ78s+eCvyqirwtbocZwdspFIaupsuxkOR52o8MVnBuJBdIz6JN7o+i2ci35TZN/NjNgRWTPavyz5SPwOPRAEoPa+hiTZhcD3+hmzTeXghClTtByeX1L5LOPp7lr92fXZx04F4H7JEbWcla5gW8KfDn6rKfAmPQ9ZJJmcXNPSwJfBupr9hZeN3ZTKehixW56zUv9f8iVahy0v+tzI/5jQfCo1aj009U7uW0C099OveSf88+2d4FAqhOo5PLa17m9VAcaFfvyb9dNun8SIU9gLh0dnFdXlqA4Xv9etN5vHsEXgOipDaP5VekGhc/gEBPkaF//1kJvNdru1CkZWUNZ2X/i+2BXsv9zMvtZwtZXgIioVqZ2ZuSdTWDMb+9N5gsjp7Y/N+eASKpVqOSt+bTNQOxu2PX70p2ZC6u+lIPAAhpVTukMwdieXVa6P9cHnFUN07ycrG27mrB6GRawST0vNST9V1VG2tiNL/fktdR+o/s+dzVQdCBaprZss/pJY2tFVvCumqYKh6U0Nr+ufZb7DDR6ikVUHmS9l7kn+o76naYncyqBiq3tzQk/xD6u6mv+Zvj5D7yWCfztPTNyefSSTqVlVtDT4dVAxVba0dSDaklqZv7jydoEfIz3RQlvkfjX+fWph+IVFX31uzrnqbt/OGFTurt9WsbehJ1KVfSN+b/WrLUXTuEbJK7VOTJzV9JXtr+tHUM8lXErWJlrq++tXV66o2VW9Z/n7V9sqd5UMVQxVS8YkiXYVUDJUPVe6s2l71fvWWqk3V6+pX1/UlWhI1qZdTz6Qfzd6auTh5UvtUrBs1/X+5Hi8nVwaXygAAAABJRU5ErkJggg=="
```
Comment (sand0) [inline EventsExpress/ClientApp/src/constants/userImage.js]: let rename to defaultImage instead

**Review by sand0 — DISMISSED** [review]
So, if we ignore all PhotoUrl when do mapping, do we use that property in any place? If not I guess we should delete it at all

File: EventsExpress.Core/Services/PhotoService.cs [hunk 3]
```diff
@@ -130,14 +131,14 @@ private async Task UploadPhotoToBlob(byte[] photo, string url)
...
 
-        public async Task<string> GetPhotoFromAzureBlob(string url)
+        public async Task<byte[]> GetPhotoFromAzureBlob(string url)
         {
             try
             {
                 using var previewMS = new MemoryStream();
                 BlobClient blobClient = _blobContainerClient.GetBlobClient(url);
                 await blobClient.DownloadToAsync(previewMS);
-                return previewMS.ToArray().ToRenderablePictureString();
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/PhotoService.cs:140]: Did we remove this extension method?

File: EventsExpress.Test/MapperTests/CommentMapperProfileTests.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ protected virtual void Init()
             IServiceProvider serviceProvider = services.BuildServiceProvider();
 
             Mapper = serviceProvider.GetService<IMapper>();
-            mock.Setup(x => x.GetPhotoFromAzureBlob(It.IsAny<string>())).Returns(Task.FromResult("test"));
+            mock.Setup(x => x.GetPhotoFromAzureBlob(It.IsAny<string>())).Returns(Task.FromResult(new byte[] { 1 }));
```
Comment (iuriikhrystiuk) [inline EventsExpress.Test/MapperTests/CommentMapperProfileTests.cs]: If we are not loading photos with models, do we need this mock anymore?

File: EventsExpress.Test/MapperTests/EventMapperProfileTests.cs [hunk 5]
```diff
@@ -320,7 +317,7 @@ protected virtual void Init()
             IServiceProvider serviceProvider = services.BuildServiceProvider();
 
             Mapper = serviceProvider.GetService<IMapper>();
-            mock.Setup(x => x.GetPhotoFromAzureBlob(It.IsAny<string>())).Returns(Task.FromResult("test"));
+            mock.Setup(x => x.GetPhotoFromAzureBlob(It.IsAny<string>())).Returns(Task.FromResult(new byte[] { 1 }));
```
Comment (iuriikhrystiuk) [inline EventsExpress.Test/MapperTests/EventMapperProfileTests.cs]: Same as above.

File: EventsExpress/ClientApp/src/actions/event/event-add-action.js [hunk 6]
```diff
@@ -45,7 +45,7 @@ export function publish_event(data) {
         let response = await api_serv.publishEvent(data);
         if (response.ok) {
             dispatch(setEventSuccess(true));
-            dispatch(getEvent(data));
+            dispatch(get_event(data.id));
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/actions/event/event-add-action.js]: I believe this was a fix of a bug. Please make sure you merged properly.

File: EventsExpress/ClientApp/src/components/event/event-item.js [hunk 7]
```diff
@@ -194,14 +198,14 @@ export default class EventCard extends Component {
                                         </IconButton>
                                     </Tooltip>
                                 </Link>
-                                <AuthComponent rolesMatch={[Roles.Admin]}>
+                                <AuthComponent rolesMatch={['Admin']}>
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/event/event-item.js]: Why is this change required?

File: EventsExpress/ClientApp/src/services/EventsExpressService.js [hunk 8]
```diff
@@ -20,6 +20,19 @@ export default class EventsExpressService {
...
     }
 
+    getPhoto = async (url, imgTagId) => {
+        const call = _url => fetch(this._baseUrl + url);
+
+        let res = await call(url);
+
+        if(res.ok){
+            let blob = await res.blob();
+            if(document.getElementById(imgTagId) !== null){
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/EventsExpressService.js]: This is not very obvious and it controls DOM outside of shadow DOM, which is not good. Can we return blob from this method and create a custom img component to apply it properly? Also, does the binding not work?

Comment (dyats) [reply EventsExpress/ClientApp/src/services/EventsExpressService.js]: It just doesn't wait for the result

**Review by iuriikhrystiuk — APPROVED** [review]

**Comment by dyats** [comment]
> So, if we ignore all PhotoUrl when do mapping, do we use that property in any place? If not I guess we should delete it at all

done.
